backend movies app

IuliiLapina opened the pull request.

IuliiLapina committed:
Добавить файл app.js

IuliiLapina committed:
Create user.js

IuliiLapina committed:
Создать схему user и movie

IuliiLapina committed:
Роуты и контроллеры users и movies

# возвращает информацию о пользователе (email и имя)
GET /users/me

# обновляет информацию о пользователе (email и имя)
PATCH /users/me

# возвращает все сохранённые пользователем фильмы
GET /movies

# создаёт фильм с переданными в теле
# country, director, duration, year, description, image, trailer, nameRU, nameEN и thumbnail, movieId
POST /movies

# удаляет сохранённый фильм по id
DELETE /movies/movieId

IuliiLapina committed:
Регистрация, авторизация, файл index для роутов

Все роуты подключены в index.js Оттуда подключаем роут в app.js, так код чище

IuliiLapina committed:
Исправить ошибки линтинга

IuliiLapina committed:
Логгер ошибок и запросов, централизованный обработчик ошибок

IuliiLapina committed:
fix: update user email

IuliiLapina committed:
fix: riut delete movie

IuliiLapina committed:
вынести валидацию в validatints.js

IuliiLapina committed:
add rate limit

IuliiLapina committed:
fix linting

IuliiLapina committed:
add cors

IuliiLapina committed:
add validation message

IuliiLapina committed:
config mongo url and secret key

IuliiLapina committed:
fix

Boortcore commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
    - [x] в проекте есть `package.json`;
    - [x] `.editorconfig`;
    - [x] `.eslintrc`, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
    - [x] `.gitignore`;
    - [x] в `.eslintrc` добавлено исключение для `_id`. Директивы `eslint-disable`, `eslint-disable-line` и `eslint-disable-next-line` запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды `npx eslint .` они должны отсутствовать. 
- [x] <!--3.64--> В разделе scripts файла `package.json` есть команды:
    - [x] `npm run start`, которая запускает сервер на `localhost:3000`;
    - [x] `npm run dev`, которая запускает сервер на `localhost:3000` с хот релоудом.
- [ ] <!--21.81--> Описанные роуты работают корректно: 
    - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
    - [x] `PATCH /users/me` — обновляет информацию о пользователе;
    - [x] `GET /movies` — все сохранённые пользователем фильмы;
    - [ ] `POST /movies` — создаёт фильм с переданными в теле данными;
_- Поле movieId должно быть числом в десятичном формате. api https://api.nomoreparties.co/beatfilm-movies возвращают фильмы, у которых id (moviId для фильмов в нашей базе данных) является числовым значением. Пожалуйста, не забудьте поменять тип поля в схеме модели и в схеме валидации celebrate._
    - [x] `DELETE /movies/:id` — удаляет сохранённый фильм по `_id`;
    - [ ] `POST /signup` — создаёт пользователя с переданными в теле данными;
_- По заданию пароль может состоять из любых символов, а не только из латинских букв и цифр. Позволять пользователям вводить только латинские буквы и цифры - это небезопасный подход. В реальных проектах, напротив, от пользователей часто требуется обязательное наличие специальных символов в пароле, чтобы его было сложнее взломать._
    - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
    - [x] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённую карточку из профиля другого пользователя. 
- [x] <!--3.64--> Все роуты, кроме `/signin` и `/signup`, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [ ] <!--7.28--> Ошибки API обрабатываются: 
    - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
    - [x] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
    - [ ] сообщение об ошибке соответствует ее типу;+
_- В случае неправильного ввода пароля или почтового ящика правильнее возвращать ошибку с текстом "Неправильная почта или пароль". Маршрут аутентификации /signin не защищён авторизацией. Поэтому ошибка с текстом "Произошла ошибка авторизации" не совсем оправдана._
    - [x] асинхронные обработчики завершены блоком `catch`;
    - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
    - [x] пароль хранится в виде хеша;
    - [x] API не возвращает хеш пароля клиенту.
- [ ] <!--7.28--> Валидация реализована верно:
    - [ ] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.
_- Попытка создать пользователя без имени не прерывается celebrate и уходит контроллеру. В схеме celebrate для всех полей необходимо указывать обязательное их наличие в запросе._
    - [ ] Данные валидируются перед добавлением в базу.
_- Попытка создать пользователя без имени должна сопровождаться ошибкой со статусом 400._
![image](https://user-images.githubusercontent.com/17438227/133655947-ccf207ab-f741-4a6b-b75c-e49a2f349b3f.png)
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из `process.env`.
- [x] <!--3.64--> К серверу можно обратиться с помощью https по адресу домена, указанному в README.md.
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT: 
    - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
    - [x] в режиме разработки (при  `process.env.NODE_ENV !== 'production'`) код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--2.14--> Все роуты подключены в файле `index.js`, который находится в папке `routes`. Оттуда единый роут подключается в файле `app.js`. 
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--2.14--> Валидация описана в отдельном модуле.
- [x] <!--2.14--> Настроено логгирование:
    - [x] запросы и ответы записываются в файл `request.log`;
    - [x] ошибки записываются в файл `error.log;`
    - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.14--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone.
    Подходящий адрес: movies-explorer.nomoreparties.co/api.
    Неподходящий адрес: movies-explorer.nomoreparties.co.

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор `Error`. 
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью. 
- [ ] <!--0.71--> Конфигурация и константы в отдельных файлах: 
    - [x] адрес Mongo-сервера, секретный ключ для JWT в режиме разработки, вынесены в отдельный конфигурационный файл;
    - [ ] константы приложения — сообщения ответов и ошибок — вынесены в отдельный файл с константами.
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено. 
- [x] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.71--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co. 
- [x] <!--0.71--> Для имени базы данных используется `moviesdb` 

## Количество баллов: 63

## Доп. Комментарии:
- В качестве валидации ссылок в схемах celebrate предлагаю использовать метод `custom` и библиотеку validator https://github.com/hapijs/joi/blob/master/API.md#anycustommethod-description 
![image](https://user-images.githubusercontent.com/17438227/83885018-732f7880-a74e-11ea-905e-28d6e8ef00d1.png)
Это избавит от необходимости писать свои регулярные выражения. 
_- При использовании `next` в функции-контроллере необходимо завершать её работу ключевым словом `return`, чтобы функция не продолжала своё выполнение._- 
![image](https://user-images.githubusercontent.com/17438227/133656242-c741fbb8-5ff6-450c-b665-05fccbbdd9e4.png)

IuliiLapina committed:
Review#1

Валидация:
- поле name обязательно
-  валидации ссылок методом custom

Роуты:
- пароль может быть любым
- Поле movieId тип number в схеме модели и валидации

Ошибки API:
- текст ошибки при авторизации с неправильным паролем/почтой

Boortcore commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
    - [x] в проекте есть `package.json`;
    - [x] `.editorconfig`;
    - [x] `.eslintrc`, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
    - [x] `.gitignore`;
    - [x] в `.eslintrc` добавлено исключение для `_id`. Директивы `eslint-disable`, `eslint-disable-line` и `eslint-disable-next-line` запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды `npx eslint .` они должны отсутствовать. 
- [x] <!--3.64--> В разделе scripts файла `package.json` есть команды:
    - [x] `npm run start`, которая запускает сервер на `localhost:3000`;
    - [x] `npm run dev`, которая запускает сервер на `localhost:3000` с хот релоудом.
- [x] <!--21.81--> Описанные роуты работают корректно: 
    - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
    - [x] `PATCH /users/me` — обновляет информацию о пользователе;
    - [x] `GET /movies` — все сохранённые пользователем фильмы;
    - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
    - [x] `DELETE /movies/:id` — удаляет сохранённый фильм по `_id`;
    - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
    - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
    - [x] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённую карточку из профиля другого пользователя. 
- [x] <!--3.64--> Все роуты, кроме `/signin` и `/signup`, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [x] <!--7.28--> Ошибки API обрабатываются: 
    - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
    - [x] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
    - [x] сообщение об ошибке соответствует ее типу;
    - [x] асинхронные обработчики завершены блоком `catch`;
    - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
    - [x] пароль хранится в виде хеша;
    - [x] API не возвращает хеш пароля клиенту.
- [x] <!--7.28--> Валидация реализована верно:
    - [x] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.
    - [x] Данные валидируются перед добавлением в базу.
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из `process.env`.
- [x] <!--3.64--> К серверу можно обратиться с помощью https по адресу домена, указанному в README.md.
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT: 
    - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
    - [x] в режиме разработки (при  `process.env.NODE_ENV !== 'production'`) код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--2.14--> Все роуты подключены в файле `index.js`, который находится в папке `routes`. Оттуда единый роут подключается в файле `app.js`. 
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--2.14--> Валидация описана в отдельном модуле.
- [x] <!--2.14--> Настроено логгирование:
    - [x] запросы и ответы записываются в файл `request.log`;
    - [x] ошибки записываются в файл `error.log;`
    - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.14--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone.
    Подходящий адрес: movies-explorer.nomoreparties.co/api.
    Неподходящий адрес: movies-explorer.nomoreparties.co.

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор `Error`. 
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью. 
- [ ] <!--0.71--> Конфигурация и константы в отдельных файлах: 
    - [x] адрес Mongo-сервера, секретный ключ для JWT в режиме разработки, вынесены в отдельный конфигурационный файл;
    - [ ] константы приложения — сообщения ответов и ошибок — вынесены в отдельный файл с константами.
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено. 
- [x] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.71--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co. 
- [x] <!--0.71--> Для имени базы данных используется `moviesdb` 

## Количество баллов: 99
